File: samples/javascript_typescript/52.enterprise-bot/src/dialogs/main/mainDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { ConversationState, StatePropertyAccessor, UserState } from "botbuilder";
import { LuisRecognizer } from "botbuilder-ai";
import { DialogContext } from "botbuilder-dialogs";
import { BotServices } from "../../botServices";
import { EscalateDialog } from "../escalate/escalateDialog";
import { OnboardingDialog } from "../onboarding/onboardingDialog";
import { OnboardingState } from "../onboarding/onboardingState";
import { RouterDialog } from "../shared/routerDialog";
import { MainResponses } from "./mainResponses";

export class MainDialog extends RouterDialog {
    private readonly _services: BotServices;
    private readonly _userState: UserState;
    private readonly _conversationState: ConversationState;
    private readonly _responder: MainResponses = new MainResponses();
    private readonly _onboardingAccessor: StatePropertyAccessor<OnboardingState>;

    constructor(services: BotServices, conversationState: ConversationState, userState: UserState) {
        super("MainDialog");
        if (!services) { throw new Error(("Missing parameter.  botServices is required")); }
        if (!conversationState) { throw new Error(("Missing parameter.  conversationState is required")); }
        if (!userState) { throw new Error(("Missing parameter.  userState is required")); }

        this._services = services;
        this._conversationState = conversationState;
        this._userState = userState;

        this._onboardingAccessor = this._userState.createProperty<OnboardingState>("OnboardingState");

        this.addDialog(new OnboardingDialog(this._services, this._onboardingAccessor));
        this.addDialog(new EscalateDialog(this._services));
    }

    protected async onStart(innerDC: DialogContext): Promise<void> {
        const onboardingState = await this._onboardingAccessor.get(innerDC.context, undefined);

        await this._responder.replyWith(innerDC.context, MainResponses.Intro);

        if (!onboardingState || !onboardingState.name) {
            // This is the first time the user is interacting with the bot, so gather onboarding information.
            await innerDC.beginDialog("OnboardingDialog");
        }
    }

    protected async route(dc: DialogContext): Promise<void> {
        // Check dispatch result
        const dispatchResult = await this._services.dispatchRecognizer.recognize(dc.context);
        const topIntent = LuisRecognizer.topIntent(dispatchResult);

        if (topIntent === "l_General") {
            // If dispatch result is general luis model
            const luisService = this._services.luisServices.get(process.env.LUIS_GENERAL || "");
            if (!luisService) { return Promise.reject(new Error("Luis service not found")); }

            const luisResult = await luisService.recognize(dc.context);

            const generalIntent = LuisRecognizer.topIntent(luisResult);

            // switch on general intents
            switch (generalIntent) {
                case "Greeting": {
                    // Send greeting response
                    await this._responder.replyWith(dc.context, MainResponses.Greeting);
                    break;
                }
                case "Help": {
                    // Send help response
                    await this._responder.replyWith(dc.context, MainResponses.Help);
                    break;
                }
                case "Cancel": {
                    // Send cancelled response.
                    await this._responder.replyWith(dc.context, MainResponses.Cancelled);

                    // Cancel any active dialogs on the stack.
                    await dc.cancelAllDialogs();
                    break;
                }
                case "Escalate": {
                    // Start escalate dialog.
                    await dc.beginDialog("EscalateDialog");
                    break;
                }
                default: {
                    // No intent was identified, send confused message.
                    await this._responder.replyWith(dc.context, MainResponses.Confused);
                    break;
                }
            }
        } else if (topIntent === "q_FAQ") {
            const qnaService = this._services.qnaServices.get("FAQ");
            if (!qnaService) { return Promise.reject(new Error("QnA service not found")); }

            const answers = await qnaService.getAnswersAsync(dc.context);

            if (answers && answers.length !== 0) {
                await dc.context.sendActivity(answers[0].answer);
            }
        }
    }

    protected complete(innerDC: DialogContext): Promise<void> {
        // The active dialogs stack ended with a complete status.
        return this._responder.replyWith(innerDC.context, MainResponses.Completed);
    }
}

File: samples/csharp_dotnetcore/52.enterprise-bot/Dialogs/Main/MainDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Luis;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;

namespace EnterpriseBot
{
    public class MainDialog : RouterDialog
    {
        private BotServices _services;
        private UserState _userState;
        private ConversationState _conversationState;
        private MainResponses _responder = new MainResponses();

        public MainDialog(BotServices services, ConversationState conversationState, UserState userState)
            : base(nameof(MainDialog))
        {
            _services = services ?? throw new ArgumentNullException(nameof(services));
            _conversationState = conversationState;
            _userState = userState;

            AddDialog(new OnboardingDialog(_services, _userState.CreateProperty<OnboardingState>(nameof(OnboardingState))));
            AddDialog(new EscalateDialog(_services));
        }

        protected override async Task OnStartAsync(DialogContext innerDc, CancellationToken cancellationToken = default(CancellationToken))
        {
            var onboardingAccessor = _userState.CreateProperty<OnboardingState>(nameof(OnboardingState));
            var onboardingState = await onboardingAccessor.GetAsync(innerDc.Context, () => new OnboardingState());

            var view = new MainResponses();
            await view.ReplyWith(innerDc.Context, MainResponses.Intro);

            if (string.IsNullOrEmpty(onboardingState.Name))
            {
                // This is the first time the user is interacting with the bot, so gather onboarding information.
                await innerDc.BeginDialogAsync(nameof(OnboardingDialog));
            }
        }

        protected override async Task RouteAsync(DialogContext dc, CancellationToken cancellationToken = default(CancellationToken))
        {
            // Check dispatch result
            var dispatchResult = await _services.DispatchRecognizer.RecognizeAsync<Dispatch>(dc.Context, CancellationToken.None);
            var intent = dispatchResult.TopIntent().intent;

            if (intent == Dispatch.Intent.l_General)
            {
                // If dispatch result is general luis model
                var luisService = _services.LuisServices["<YOUR MS BOT NAME>_General"];
                var result = await luisService.RecognizeAsync<General>(dc.Context, CancellationToken.None);

                var generalIntent = result?.TopIntent().intent;

                // switch on general intents
                switch (generalIntent)
                {
                    case General.Intent.Greeting:
                        {
                            // send greeting response
                            await _responder.ReplyWith(dc.Context, MainResponses.Greeting);
                            break;
                        }

                    case General.Intent.Help:
                        {
                            // send help response
                            await _responder.ReplyWith(dc.Context, MainResponses.Help);
                            break;
                        }

                    case General.Intent.Cancel:
                        {
                            // Send cancelled response.
                            await _responder.ReplyWith(dc.Context, MainResponses.Cancelled);

                            // Cancel any active dialogs on the stack.
                            await dc.CancelAllDialogsAsync();
                            break;
                        }

                    case General.Intent.Escalate:
                        {
                            // Start escalate dialog.
                            await dc.BeginDialogAsync(nameof(EscalateDialog));
                            break;
                        }

                    case General.Intent.None:
                    default:
                        {
                            // No intent was identified, send confused message.
                            await _responder.ReplyWith(dc.Context, MainResponses.Confused);
                            break;
                        }
                }
            }
            else if (intent == Dispatch.Intent.q_FAQ)
            {
                var qnaService = _services.QnAServices["<YOUR_QNA_SERVICE_NAME>"];
                var answers = await qnaService.GetAnswersAsync(dc.Context);

                if (answers != null && answers.Count() > 0)
                {
                    await dc.Context.SendActivityAsync(answers[0].Answer);
                }
            }
        }

        protected override async Task CompleteAsync(DialogContext innerDc, CancellationToken cancellationToken = default(CancellationToken))
        {
            // The active dialogs stack ended with a complete status.
            await _responder.ReplyWith(innerDc.Context, MainResponses.Completed);
        }
    }
}
